exchange/bitswap/session.go

```diff
@@ -0,0 +1,233 @@
+package bitswap
+
+import (
+	"context"
+	"time"
+
+	blocks "github.com/ipfs/go-ipfs/blocks"
+	notifications "github.com/ipfs/go-ipfs/exchange/bitswap/notifications"
+
+	logging "gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log"
+	lru "gx/ipfs/QmVYxfoJQiZijTgPNHCHgHELvQpbsJNTg6Crmc3dQkj3yy/golang-lru"
+	loggables "gx/ipfs/QmVesPmqbPp7xRGyY96tnBwzDtVV1nqv4SCVxo5zCqKyH8/go-libp2p-loggables"
+	cid "gx/ipfs/QmYhQaCYEcaPPjxJX7YcPcVKkQfRy6sJ7B3XmGFk82XYdQ/go-cid"
+	peer "gx/ipfs/QmdS9KpbDyPrieswibZhkod1oXqRwZJrUPzxCofAMWpFGq/go-libp2p-peer"
+)
+
+const activeWantsLimit = 16
+
+// Session holds state for an individual bitswap transfer operation.
+// This allows bitswap to make smarter decisions about who to send wantlist
+// info to, and who to request blocks from
+type Session struct {
+	ctx            context.Context
+	tofetch        []*cid.Cid
+	activePeers    map[peer.ID]struct{}
+	activePeersArr []peer.ID
+
+	bs         *Bitswap
+	incoming   chan blkRecv
+	newReqs    chan []*cid.Cid
+	cancelKeys chan []*cid.Cid
+
+	interest  *lru.Cache
```

Comment:
no, this is used to track blocks that are relevant to this session, so responses from other sessions can inform us of peers. Even after we've received a block, it remains in the interest cache, because receiving a block from peer X a different session may mean that peer X also has other blocks our session wants.